### ISSUE TITLE
Add a code action on imports to add/remove aliases

### ISSUE THREAD

**Issue by ValentinVignal** [Dart-Code/Dart-Code#4289]
**Is your feature request related to a problem? Please describe.**

My personal use case is migrating from [`provider`](https://pub.dev/packages/provider) to [`riverpod`](https://pub.dev/packages/riverpod). I'm introducing the import `flutter_riverpod` in files where `provider` is already imported. But both libraries are exporting a type `ChangeNotifierProvider`.
I'm migrating from `provider` to `riverpod` incrementally, so for now, I'm using both libraries. But ultimately, I will only use `riverpod`. I would be very useful to be able to change
```dart
import 'package:provider/provider.dart' ;
```

into 

```dart
import 'package:provider/provider.dart'  as provider;
```

and update the file accordingly.

**Describe the solution you'd like**

It would be nice to add an option "add/remove alias` to an import which adds/removes the alias and update the rest of the file accordingly.

![image](https://user-images.githubusercontent.com/32538273/204721766-093b79cc-839b-4862-a697-da2c2e4e9abc.png)

**Describe alternatives you've considered**

I'm updating the file by hand.

**Comment by DanTup** [comment]
The server does have some existing support for this, where you can invoke "rename" on an `import` to change (or add/remove) its alias). However, there are some issues with VS Code's handling of this that make the experience not work correctly (see https://github.com/microsoft/vscode/issues/163929).

That said, even if fixed, performing a rename in this way might not be the most obvious way to add/remove an alias, so I wonder if having it in code actions would be beneficial anyway.

@bwilkerson thoughts? This could be done via the new refactoring mechanism (to prompt for a name when adding/renaming) and probably just wired up to the existing rename behaviour.

**Comment by bwilkerson** [comment]
I've always thought that it was a bit odd to have a "rename" that renames to or from nothing (or an empty string). It's always seemed to me that it would be more clear to users if we had an "Add Prefix" and "Remove Prefix" pair of operations for to handle those cases (with "Rename" still handling the case of changing one prefix into a different, but non-empty, prefix).

I think the new refactoring mechanism gives us an opportunity to do just that.

**Comment by FMorschel** [comment]
Today by triggering a rename on the `import` keyword we already can rename/add an alias/prefix to the import and all the elements that are exported by it.

![image](https://github.com/user-attachments/assets/6d835956-29e6-458f-af19-ececcbcfcccb)
![image](https://github.com/user-attachments/assets/41f7b52a-e7a6-42fb-852c-6b79b02ef158)
![image](https://github.com/user-attachments/assets/58b1a74a-b08a-4e07-958e-eea977162d14)

From @bwilkerson's comment here: https://github.com/dart-lang/sdk/issues/55863#issuecomment-2137564338

---

We could add an assist to rename and another to remove the alias. As was discussed on Discord, here is the message from @DanTup:

> I think if we added an assist, it would prompt for the prefix, and then just do the same thing as the current rename. The difference would be that we could control what's prefilled in the input box (and we could even suggest a name). Of course, it'd be nicer if it was something native to VS Code/LSP, but that seems unlikely (I don't think VS Code likes this use case, where you're renaming something you can't currently see in the editor).

I like the idea of an assist that would suggest the name of the package or something similar like the first letters for package words like `device_info_plus` becoming `dip` or something.

**Comment by bwilkerson** [comment]
Would it really be an improvement to have two different ways of performing what is essentially the same operation? That's something we would usually not do. (Also, the burden of maintaining both implementations is a bit of a negative.)

I'd want to see some evidence that a large percentage of users want this before we add it.

**Comment by FMorschel** [comment]
I personally only would like the assist for better discoverability since it is weird to "rename an import" even more so when it shows the `import` keyword to begin the action making people wonder if that is actually what they want to do (I got a bit scared the first time using it). 

But if we had something documenting this kind of actions that would not be necessary IMO.

But the remove one is needed since today we can't remove the alias by clearing the name box that appears because VsCode doesn't handle empty strings as a valid "rename".

**Comment by bwilkerson** [comment]
> ... VsCode doesn't handle empty strings as a valid "rename".

That's unfortunate. IntelliJ handles it (or did the last time I checked).

> ... better discoverability ...

That's a valid point, and makes me think maybe we ought to add them. It would be good to get more feedback before doing so, however.

**Comment by DanTup** [comment]
Yeah, the VS Code issue is tracked at https://github.com/microsoft/vscode/issues/163929

The issue is that if we provide an empty range for the source being renamed, VS Code treats it like we couldn't compute it and it uses the word under the cursor:

https://github.com/user-attachments/assets/1976243b-d6a1-4ef4-9f5a-3f165054ae12

Ofc, it's a shame to have to do things differently because of a specific editors quirks, but I do think this feature is not very discoverable to users even without that bug.

**Comment by FMorschel** [comment]
> > ... VsCode doesn't handle empty strings as a valid "rename".
> 
> That's unfortunate. IntelliJ handles it (or did the last time I checked).
> 
> > ... better discoverability ...
> 
> That's a valid point, and makes me think maybe we ought to add them. 

I was thinking about this a bit more this week, I think adding the same for named constructors could be nice too. As stated above, VS Code doesn't handle removing the name entirely, and also, renaming on the implicit `new` constructor is a bit counter-intuitive for new users that are not aware of this behaviour. When I first started using Dart, the first time I tried renaming on an implicit `new` constructor, I was actually trying to rename the class. 

> It would be good to get more feedback before doing so, however.

Should we add the "your feedback wanted!" tag here?

**Comment by ValentinVignal** [comment]
Just saw the tag "your feedback wanted"

I understand why `new` could be confusing for beginners, but it seems fair enough from the limitations described.

However, we wouldn't be able to use the same kind of trick to remove the alias of an import, right?

**Comment by FMorschel** [comment]
> However, we wouldn't be able to use the same kind of trick to remove the alias of an import, right?

Currently, none is implemented. My comment above was asking that we do this for both the import alias/prefix and for constructor names. Since they are currently facing the same VS Code issue.

**Comment by ValentinVignal** [comment]
I was wondering, would it work to accept a space` ` ?

**Comment by DanTup** [comment]
@bwilkerson now that we have add/remove constructor name, do you think it makes sense to add this in the same way? (I think it'll be mostly the same, with the existing rename handling the bulk of the work?)

**Comment by bwilkerson** [comment]
Yes. I definitely think we should add the same add/remove pair for import prefixes. And yes, the implementation of those actions should look very similar to the pair for a constructor name. I've already updated the 'rename import prefix' class to use a change builder, so it should be easy to reuse.

And after we get IntelliJ to use the new code actions, we can then disable the weird behavior in 'rename' and not have two ways of doing the same thing.

**Comment by DanTup** [comment]
sgtm, I'll take a look - thanks!

Edit: https://dart-review.googlesource.com/c/sdk/+/498700

**Comment by DanTup** [comment]
Fixed by https://github.com/dart-lang/sdk/commit/ac8b7adbaf7adf5d82afc9957319207f514c6fb0, ships in a future SDK.
